### PR TITLE
[RLlib] RLlib contrib

### PR DESCRIPTION
**Pull request by avnishn**
Signed-off-by: Avnish <avnishnarayan@gmail.com>

RLlib-Contrib is a directory for more experimental community contributions to RLlib including contributed algorithms. **This directory has a more relaxed bar for contributions than ray or RLlib.**

this pr introduces RLlib with 2 algorithms as well as its very on CI [WIP] 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib_contrib/README.md [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+RLlib-Contrib is a directory for more experimental community contributions to RLlib including contributed algorithms. **This directory has a more relaxed bar for contributions than ray or RLlib.** If you are interested in contributing to RLlib-Contrib, please see the [contributing guide](CONTRIBUTING.md).
+
+## Getting Started
+Navigate to the algorithm directory you are interested in and see the README.md for installation instructions and example scripts to help you get started!
+
+## Maintenance
+
+**Any issues that are filed in `rllib_contrib` will be solved best-effort by the community and there is no expectation of maintenance by the RLlib team.**
+
+**The api surface between algorithms in rllib_contrib and current versions of ray / rllib is not guaranteed. This means that any apis that are used in rllib_contrib could potentially become modified/removed in newer version of ray/rllib.**
```
Comment (sven1977) [inline rllib_contrib/README.md:12]: Please consistently fix spelling (upper-case) of `Ray`, `RLlib`, and `APIs`.

File: rllib_contrib/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+We will generally accept contributions to this directory that meet any of the following criteria:
+
+1. Updating dependencies.
+2. Submitting community contributed algorithms that have been tested and are ready for use.
+3. Enabling algorithms to be run in different environments (ex. adding support for a new type of gym environment).
```
Comment (sven1977) [inline rllib_contrib/README.md:18]: gym -> gymnasium

File: rllib_contrib/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+4. Updating algorithms for use with the newer RLlib APIs.
+5. General bug fixes.
```
Comment (sven1977) [inline rllib_contrib/README.md:20]: Awesome!

File: rllib_contrib/a3c/README.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+conda create -n rllib-a3c python=3.10
+conda activate rllib-a3c
+python -m pip install ray==2.3.1
+python -m pip install -e '.[development]'
+pre-commit install
+```
+
+## Usage
+
+```python
```
Comment (sven1977) [inline rllib_contrib/a3c/README.md:18]: Is there any way we can simply link into the below example script via sphinx tags?

File: rllib_contrib/a3c/src/rllib_a3c/a3c/README.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# Asynchronous Advantage Actor-Critic (A3C)
```
Comment (sven1977) [inline rllib_contrib/a3c/src/rllib_a3c/a3c/README.md:1]: Here, too, seems duplicate. Should we just have one README.md in the top-level directory (`ray/rllib_contrib/a3c/README.md`)?

Comment (sven1977) [reply rllib_contrib/a3c/src/rllib_a3c/a3c/README.md:1]: Also, why is it: `scr/rllib_a3c/a3c`, but then `src/rllib_maml` (<- w/o the extra `maml` sub-dir)?

**Review by sven1977 — COMMENTED** [review]
Looks awesome. I love the modular design of this whole thing. Great initiative @avnishn , this is going to simplify RLlib so much for the better!
Just a handful of nits/questions. Otherwise, good to go.

**Comment by kouroshHakha** [comment]
Loving this ❤️ 

image

File: rllib_contrib/a3c/README.md (hunk 4 above)
Comment (kouroshHakha) [reply rllib_contrib/a3c/README.md:18]: +1

File: rllib_contrib/a3c/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,2 @@
+tensorflow==2.11.0
```
Comment (kouroshHakha) [inline rllib_contrib/a3c/requirements.txt:1]: when do we pin the ray versions?

File: rllib_contrib/README.md
Comment (kouroshHakha) [inline rllib_contrib/README.md:1]: There should be a part about installation and using these algorithms

File: rllib_contrib/a3c/src/rllib_a3c/a3c/README.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+In A3C, the worker nodes generate data asynchronously and compute gradients from the data. These
+computed gradients are then sent to the central node.
+Note that the workers in A3C may be slightly out-of-sync with the central node due to asynchrony,
+which may induce biases in learning (on-policy loss function).
+
+
+## Documentation & Implementation of A3C:
+
+**[Detailed Documentation](https://docs.ray.io/en/master/rllib-algorithms.html#a3c)**
```
Comment (kouroshHakha) [inline rllib_contrib/a3c/src/rllib_a3c/a3c/README.md:19]: these should refer to latest

**Review by kouroshHakha — COMMENTED** [review]
Awesome. It's missing docs on how to install packages and use the new rllib_contrib. We should also create a follow up PR that in the main ray repo creates deprecation notices and also links the REP on to the algorithm page and creates a note about moving the algorithms to rllib_contrib.

File: rllib_contrib/a3c/src/rllib_a3c/a3c/README.md (hunk 7 above)
Comment (avnishn) [reply rllib_contrib/a3c/src/rllib_a3c/a3c/README.md:19]: like the latest ray release -- not master. gotcha

File: rllib_contrib/README.md
Comment (avnishn) [reply rllib_contrib/README.md:1]: I renamed getting started to getting started and installation

File: rllib_contrib/a3c/requirements.txt (hunk 6 above)
Comment (avnishn) [reply rllib_contrib/a3c/requirements.txt:1]: inside of the pyproject.toml

File: rllib_contrib/a3c/README.rst [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+Installation
+------------
+
+.. code-block:: bash
+
+    conda create -n rllib-a3c python=3.10
+    conda activate rllib-a3c
+    pip install -r requirements.txt
+    pip install -e '.[development]'
```
Comment (kouroshHakha) [inline rllib_contrib/a3c/README.rst:15]: some thought:
better UX: some people may not want to clone the repos.
AI (action item): add an alternative for installation that does the cloning behind the scene. 

```
conda create -n rllib-a3c python=3.10
conda activate rllib-a3c
pip install -r https://raw.githubusercontent.com/ray-project/ray/master/rllib/contrib/a3c/requirements.txt
pip install git+https://github.com/ray-project/ray.git#subdirectory=rllib_contrib/a3c
```

**Review by kouroshHakha — COMMENTED** [review]
Looks good. one nit which we can cover in another PR or later. so good to merge as is. Just let me know when you want me to push the button.

File: rllib_contrib/a3c/README.rst (hunk 8 above)
Comment (avnishn) [reply rllib_contrib/a3c/README.rst:15]: I've added this for now, and we may need a separate pr to fix the links after we merge this pr.

Comment (avnishn) [reply rllib_contrib/a3c/README.rst:15]: actually I'm going to open a separate pr for this lets merge as is

**Review by kouroshHakha — APPROVED** [review]

**Pull request merged**